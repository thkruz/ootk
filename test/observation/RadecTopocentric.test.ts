import { DEG2RAD, Degrees, EpochUTC, J2000, Kilometers, KilometersPerSecond, RadecTopocentric,
  Radians, RadiansPerSecond, Vector3D } from '../../src/main';
describe('RadecTopocentric', () => {
  let radec: RadecTopocentric;
  let exampleDate: Date;

  beforeEach(() => {
    exampleDate = new Date(1705109326817);
    radec = new RadecTopocentric(
      EpochUTC.fromDateTime(exampleDate),
      25 * DEG2RAD as Radians,
      0 * DEG2RAD as Radians,
    );
  });

  it('should be constructable', () => {
    expect(radec).toMatchSnapshot();
  });

  // fromDegrees
  it('should be constructable from degrees', () => {
    const radecDeg = RadecTopocentric.fromDegrees(
      EpochUTC.fromDateTime(exampleDate),
      25 as Degrees,
      0 as Degrees,
    );

    expect(radecDeg).toMatchSnapshot();
  });

  // fromStateVector
  it('should be constructable from a state vector', () => {
    const site = new J2000(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(8000 as Kilometers, 7000 as Kilometers, 8000 as Kilometers),
      new Vector3D<KilometersPerSecond>(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );
    const state = new J2000(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(7000 as Kilometers, 7000 as Kilometers, 8000 as Kilometers),
      new Vector3D<KilometersPerSecond>(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );
    const radecSv = RadecTopocentric.fromStateVector(state, site);

    expect(radecSv).toMatchSnapshot();
  });

  // rightAscensionDegrees
  it('should return the right ascension in degrees', () => {
    expect(radec.rightAscensionDegrees).toMatchSnapshot();
  });

  // declinationDegrees
  it('should return the declination in degrees', () => {
    expect(radec.declinationDegrees).toMatchSnapshot();
  });

  // rightAscensionRateDegrees
  it('should return the right ascension rate in degrees', () => {
    expect(radec.rightAscensionRateDegrees).toMatchSnapshot();
  });

  // declinationRateDegrees
  it('should return the declination rate in degrees', () => {
    expect(radec.declinationRateDegrees).toMatchSnapshot();
  });

  // position
  it('should return the position', () => {
    const site = new J2000(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(8000 as Kilometers, 7000 as Kilometers, 8000 as Kilometers),
      new Vector3D<KilometersPerSecond>(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(radec.position(site)).toMatchSnapshot();
  });

  // velocity
  it('should return the velocity', () => {
    const radec2 = new RadecTopocentric(
      EpochUTC.fromDateTime(exampleDate),
      25 * DEG2RAD as Radians,
      1 * DEG2RAD as Radians,
      1000 as Kilometers,
      1.5 * DEG2RAD as RadiansPerSecond,
      2.5 * DEG2RAD as RadiansPerSecond,
    );
    const site = new J2000(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(8000 as Kilometers, 7000 as Kilometers, 8000 as Kilometers),
      new Vector3D<KilometersPerSecond>(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(radec2.velocity(site)).toMatchSnapshot();
  });

  // angle
  it('should return the angle', () => {
    const radec2 = new RadecTopocentric(
      EpochUTC.fromDateTime(exampleDate),
      25 * DEG2RAD as Radians,
      1 * DEG2RAD as Radians,
    );

    expect(radec.angle(radec2)).toMatchSnapshot();
  });

  // angleDegrees
  it('should return the angle in degrees', () => {
    const radec2 = new RadecTopocentric(
      EpochUTC.fromDateTime(exampleDate),
      25 * DEG2RAD as Radians,
      1 * DEG2RAD as Radians,
    );

    expect(radec.angleDegrees(radec2)).toMatchSnapshot();
  });
});
