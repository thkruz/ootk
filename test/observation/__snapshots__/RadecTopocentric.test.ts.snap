// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadecTopocentric should be constructable 1`] = `
RadecTopocentric {
  "declination": 0,
  "declinationRate": undefined,
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "range": undefined,
  "rangeRate": undefined,
  "rightAscension": 0.4363323129985824,
  "rightAscensionRate": undefined,
}
`;

exports[`RadecTopocentric should be constructable from a state vector 1`] = `
RadecTopocentric {
  "declination": 0,
  "declinationRate": 0,
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "range": 1000,
  "rangeRate": 0,
  "rightAscension": 3.141592653589793,
  "rightAscensionRate": -0,
}
`;

exports[`RadecTopocentric should be constructable from degrees 1`] = `
RadecTopocentric {
  "declination": 0,
  "declinationRate": null,
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "range": undefined,
  "rangeRate": undefined,
  "rightAscension": 0.4363323129985824,
  "rightAscensionRate": null,
}
`;

exports[`RadecTopocentric should return the angle 1`] = `0.017453292519941554`;

exports[`RadecTopocentric should return the angle in degrees 1`] = `0.9999999999999002`;

exports[`RadecTopocentric should return the declination in degrees 1`] = `0`;

exports[`RadecTopocentric should return the declination rate in degrees 1`] = `null`;

exports[`RadecTopocentric should return the position 1`] = `
Vector3D {
  "x": 8000.906307787037,
  "y": 7000.42261826174,
  "z": 8000,
}
`;

exports[`RadecTopocentric should return the right ascension in degrees 1`] = `25`;

exports[`RadecTopocentric should return the right ascension rate in degrees 1`] = `null`;

exports[`RadecTopocentric should return the velocity 1`] = `
Vector3D {
  "x": -11.752592909334684,
  "y": 23.401642759234754,
  "z": 43.62658574738897,
}
`;
